File: DailyLeetcoding/20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for(char c : s.toCharArray()){
            switch(c){
                case '(': 
                    stack.push(')');
                    break;
                case '{': 
                    stack.push('}');
                    break;
                case '[': 
                    stack.push(']');
                    break;
                default:
                    if (stack.isEmpty() || stack.pop() != c) {
                        return false;
                    }
                    break;
            }
        }
         return stack.isEmpty();
    }
}

File: tree/main/DailyLeetcoding/20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        p = {
            ")": "(",
            "}": "{",
            "]": "["
        }

        stack = []

        for ch in s:
            if ch in p:
                if stack and stack[-1] == p[ch]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(ch)

        return not stack
        